keps/sig-apps/1847-autoremove-statefulset-pvcs/README.md

```diff
@@ -0,0 +1,508 @@
+will more durable than ephemeral volumes would be, as they are only deleted on
+scale-down or StatefulSet deletion, and not on other pod lifecycle events like
+being rescheduled to a new node, even with the new retain policies.
+
+## Design Details
+
+### Volume delete policy for the StatefulSet created PVCs
+
+A new field named `PersistentVolumeClaimDeletePolicy` of the type
+`StatefulSetPersistentVolumeClaimDeletePolicy` will be added to the StatefulSet. This
+field will represent the user indication on whether the associated PVCs can be
+automatically deleted or not. The default policy would be `Retain`.
+
+In alpha release we intend to keep the `PersistentVolumeClaimDeletePolicy` immutable after
+creation.  Based on user feedback we will consider making this field mutable in future
+releases.
+
+When a StatefulSet spec has a `VolumeClaimTemplate`, PVCs are dynamically created using a
+static naming scheme, and each Pod is created with a claim to the corresponding PVC. These
+are the precise PVCs meant when referring to the volume or PVC for Pod below, and these
+are the only PVCs modified with an ownerRef. Other PVCs referenced by the StatefulSet Pod
+template are not affected by this behavior.
+
+#### `DeleteOnScaledown`
+
+If `PersistentVolumeClaimDeletePolicy` is set to `DeleteOnScaledown`, the Pod will be set
+as the owner of the PVCs created from the `VolumeClaimTemplates` just before the
+scale-down is performed by the StatefulSet controller.  When a Pod is deleted, the PVC
+owned by the Pod is also deleted. When `DeleteOnScaledown` policy is set and the
+Statefulset gets deleted the PVCs also will get deleted (similar to
+`DeleteOnStatefulSetDeletion` policy).
+
+The current StatefulSet controller implementation ensures that the manually deleted pods
+are restored before the scale-down logic is run. This combined with the fact that the
+owner references are set only before the scale-down will ensure that manual deletions do
+not automatically delete the PVCs in question.
+
+During scale-up, if a PVC has an OwnerRef that does not match the Pod, it indicates that
+the PVC was referred to by the deleted Pod and is in the process of getting
+deleted. The controller will skip the reconcile loop until PVC deletion finishes, avoiding
+a race condition.
+
+In addition, on PVC creation an OwnerRef is added for when the StatefulSet is
+deleted. See the `DeleteOnStatefulSetDeletion` policy below for further details
+how this will be handled.
+
+#### `DeleteOnStatefulSetDeletion`
+
+When `PersistentVolumeClaimDeletePolicy` is set to
+`DeleteOnStatefulSetDeletion`, when a VolumeClaimTemplate PVC is created, an
+owner reference in PVC will be added to point to the StatefulSet. When a
+scale-up or scale-down occurs, the PVC is unchanged.  PVCs previously in use
+before scale-down will be used again when the scale-up occurs.
+
+Note that the ownerRef will be set even when a VolumeClaimTemplate PVC
```

Comment:
For an immutable policy, and setting ownerRef on creation, I don't think you could get into this state from a controller crash. You would only get here if the user manually created the PVC object.

I do think that it's better to be reconciling to support the manually created PVCs case.

Reply:
Yes, you're correct, thanks for pointing that out. I've clarified how reconcilation and ownerRefs work.